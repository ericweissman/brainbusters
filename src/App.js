import React, { Component } from 'react';
import Header from './Header';
import CardContainer from './CardContainer';
import './styles/Main.scss';

class App extends Component {
  constructor() {
    super();

    this.state = {
      questions: [],
      studyList: [],
      showStudyList: false,
      showAllQuestions: false,
      guessedQuestions: [],
    }
  }
  componentDidMount() {
    fetch("http://memoize-datasets.herokuapp.com/api/v1/ewQuestions")
      .then(results => results.json())
      .then((result) => {
        this.setState({
          questions: result.ewQuestions
        })
      })
      .catch(err => {
        this.setState({ error: err })
      })

      this.populateStudyList()
    }

  populateStudyList = () => {
    if (Object.keys(localStorage).length > 0) {
      const savedStudyList = JSON.parse(localStorage.getItem('StudyList'))
      this.setState({
        studyList: savedStudyList
      })
    }
  }

  toggleAllQuestions = () => {
    this.setState({
      showAllQuestions: !this.state.showAllQuestions
    })
  }

  updateStudyList = (id) => {
    const questions = [...this.state.questions];
    const studyList = [...this.state.studyList];
    const missedQuestions = questions.find((question) => {
      return question.id === id
    })
    if (!studyList.includes(missedQuestions.id)) {
      studyList.push(missedQuestions.id)
    }
    this.setState({
      studyList: studyList
    }, () => {
      localStorage.setItem('StudyList', JSON.stringify(studyList))
      this.updateGuessedCards(id);
    })

    
  }

  updateGuessedCards = (id) => {
    const questionsArr = [...this.state.questions];
    const guessedArr = [...this.state.guessedQuestions];
    const guessed = questionsArr.find((elem) => {
      return elem.id === id
    })
    if (!guessedArr.includes(guessed.id)) {
      guessedArr.push(guessed.id)
    }
    this.setState({
      guessedQuestions: guessedArr
    })

  }

  render() {
      return (
        <div className="App">
          <Header
            toggle={this.toggleAllQuestions}
          />
          <CardContainer
            questions={this.state.questions}
            updateStudyList={this.updateStudyList}
            studyList={this.state.studyList}
            showAllQuestions={this.state.showAllQuestions}
            updateGuessedCards={this.updateGuessedCards}
            guessedQuestions={this.state.guessedQuestions}
          />
        </div>
      );
    }
  }

export default App;
