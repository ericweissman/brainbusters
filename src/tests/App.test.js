import React from 'react';
import ReactDOM from 'react-dom';
import App from '../App';
import { shallow, mount } from 'enzyme';


const question = [
  {
    "id": 1,
    "question": "What array prototype method would return a sum of the elements in an array?",
    "answers": [
      ".sort()",
      ".reduce()",
      ".forEach()"
    ],
    "correct_answer": ".reduce()",
    "concept": "prototypes"
  }];
const guessedQuestions = [];
const id = 1;
const studyList = [4, 6]

describe('App', () => {
  let wrapper;

  beforeEach(() => {
    // localStorage.setItem('StudyList', '[4, 6]');
    wrapper = shallow(
      <App />
    );

  afterEach(() => {
    localStorage.clear();
  })

  });

  it('should retrieve data from local storage on mount', () => {
    localStorage.setItem('StudyList', JSON.stringify(studyList))
    wrapper = mount(<App />)
    expect(wrapper.state().studyList).toEqual(studyList)
  }) 

  it('should only update Guessed Cards with a new ID when the id matches an ID in guessed questions', () => {
    wrapper.setState({ questions: question, guessedQuestions: guessedQuestions});
    wrapper.instance().updateGuessedCards(id);
    expect(wrapper.state('guessedQuestions')).toEqual([1]);
    wrapper.instance().updateGuessedCards(id);
    expect(wrapper.state('guessedQuestions')).toEqual([1]);
  });



  it('should match the snapshot with all data passed in correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render the Header and CardContainer component', () => {
    expect(wrapper.find('Header').length).toEqual(1)
    expect(wrapper.find('CardContainer').length).toEqual(1)
  });



  it('should have a default state', () => {
    expect(wrapper.state()).toEqual({
      questions: [],
      studyList: [],
      showAllQuestions: false,
      guessedQuestions: [],
      error: null
    })
  });

  it('should change state when toggleAllQuestions is called', () => {
    expect(wrapper.state('showAllQuestions')).toEqual(false);
    wrapper.instance().toggleAllQuestions();
    expect(wrapper.state('showAllQuestions')).toEqual(true);
    wrapper.instance().toggleAllQuestions();
    expect(wrapper.state('showAllQuestions')).toEqual(false);
  });

  it('should change state when resetQuiz is called', () => {
    wrapper.setState({
      questions: [1,2],
      studyList: [1,2],
      showAllQuestions: false,
      guessedQuestions: []
    });
    wrapper.instance().resetQuiz();
    expect(wrapper.state()).toEqual({
      questions: [],
      studyList: [],
      showAllQuestions: false,
      guessedQuestions: []
    });
  })

  
});