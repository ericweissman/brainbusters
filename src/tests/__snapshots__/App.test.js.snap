// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot with all data passed in correctly 1`] = `
<div
  className="App"
>
  <div
    className="app"
  >
    <Header
      resetQuiz={[Function]}
      showAllQuestions={false}
      studyList={Array []}
      toggle={[Function]}
    />
    <CardContainer
      guessedQuestions={Array []}
      questions={Array []}
      showAllQuestions={false}
      studyList={Array []}
      updateGuessedCards={[Function]}
      updateStudyList={[Function]}
    />
  </div>
</div>
`;
